consul/consul.go

```diff
@@ -137,15 +134,15 @@ func (c *Consul) Register(service *registry.Service) {
 //
 func (c *Consul) Deregister() error {
 	for s, b := range serviceCache {
-		if !b.isRegistered {
+		if c.CacheIsValid(s) {
+			c.CacheProcessDeregister(s)
+		} else {
```

Comment:
Now that I think about it maybe the flag should be --failed-heartbeats-before-remove?